106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+
+```mermaid
+sequenceDiagram
+  autonumber
+  RebalanceOnScaleUp->>RebalanceOnScaleDown: queued scale down
+  RebalanceOnScaleUp->>RebalanceOnScaleUp: queued scale up
+  RebalanceOnScaleUp->>RebalanceOnAnomalyDetection: queued skew rebalance
+  critical queued scale up and skew rebalance
+    RebalanceOnScaleUp->>RebalanceOnScaleUp: execute scale down
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  critical queued scale up and scale down
+    RebalanceOnScaleUp->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleUp-->>RebalanceOnScaleUp: then scale up
+  end
+  critical queued scale down and scale up and skew rebalance
+    RebalanceOnScaleUp->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  RebalanceOnScaleUp->>Idle: scale up complete with no queued items
+```
+
+* from **RebalanceOnScaleUp**:
+  * **RebalanceOnScaleUp**: if a rebalancing on scale up is still running or another one was requested while the first one ended.
+  * **RebalanceOnScaleDown**: if a scale down operation was requested, so the current rebalancing scale up is stopped (and queued) and a new rebalancing scale down is started. The rebalancing scale up will be postponed.
+  * **RebalanceOnAnomalyDetection**: if a configmap related to goal violation was detected. It will run once the queued scale down and scale up is completed.
+  * **Idle**: if a scale up operation was requested, it was executed and completed successfully or failed.
+
+```mermaid
+sequenceDiagram
+  autonumber
+  RebalanceOnAnomalyDetection->>RebalanceOnScaleDown: queued scale down
+  RebalanceOnAnomalyDetection->>RebalanceOnScaleUp: queued scale up
+  RebalanceOnAnomalyDetection->>RebalanceOnAnomalyDetection: queued skew rebalance
+  critical queued scale up and skew rebalance
+    RebalanceOnAnomalyDetection->>RebalanceOnScaleUp: execute scale up
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  critical queued scale down and skew rebalance
+    RebalanceOnAnomalyDetection->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  critical queued scale down and scale up and skew rebalance
+    RebalanceOnAnomalyDetection->>RebalanceOnScaleDown: execute scale down
+    RebalanceOnScaleDown-->>RebalanceOnScaleUp: then scale up
+    RebalanceOnScaleUp-->>RebalanceOnAnomalyDetection: then skew rebalance
+  end
+  RebalanceOnAnomalyDetection->>Idle: scale up complete with no queued items
+```
+* from **RebalanceOnAnomalyDetection**:
+  * **RebalanceOnAnomalyDetection**: if another goal violation was detected while the first one ended. If a scale down and scale up is also queued up then they will execute first.
```

Comment:
I don't get what this means? Where are things "queued up"? Where is this queue?

Reply:
Also your bottom point here says that if a scale down is requested then it will wait for the skew to finish. That seems to contradict what you say here? Or do you mean if a _new_ ROAD is posted that will go after a scale up/down?

Reply:
Yes, I meant if a new ROAD is posted then that will go after a scale up/down?